### ISSUE TITLE
Planner bug: wrong result if sub-selects contain volatiles

### ISSUE THREAD

**Issue by hsyuan** [greenplum-db/gpdb#2608]
This is query from upstream.

Repro:
```sql
select x, x from (select (select random()) as x from (values(1),(2)) v(y)) ss;
         x         |        x
-------------------+------------------
 0.679719430860132 | 0.26834974065423
 0.679719430860132 | 0.26834974065423
(2 rows)

explain select x, x from (select (select random()) as x from (values(1),(2)) v(y)) ss;
                         QUERY PLAN
-------------------------------------------------------------
 Values Scan on "*VALUES*"  (cost=0.03..0.05 rows=1 width=0)
   InitPlan
     ->  Result  (cost=0.00..0.01 rows=1 width=0)
   InitPlan
     ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Optimizer status: legacy query optimizer
(6 rows)
```

The 2 `x` should have the same value.

**Comment by hsyuan** [comment]
Upstream generates the following plan, which seems more reasonable:
```sql
explain select x, x from (select (select random()) as x from (values(1),(2)) v(y)) ss;
                            QUERY PLAN
-------------------------------------------------------------------
 Subquery Scan on ss  (cost=0.01..0.06 rows=2 width=16)
   ->  Values Scan on "*VALUES*"  (cost=0.01..0.04 rows=2 width=8)
         InitPlan 1 (returns $0)
           ->  Result  (cost=0.00..0.01 rows=1 width=8)
(4 rows)
```

**Comment by hlinnaka** [comment]
PostgreSQL 8.3 actually produces the same, incorrect, result, if you tweak the query a little bit to prevent to innermost subquery from being pulled-up:
```
postgres=# select version();
                                              version                                               
----------------------------------------------------------------------------------------------------
 PostgreSQL 8.3.23 on x86_64-unknown-linux-gnu, compiled by GCC gcc (Debian 6.2.1-5) 6.2.1 20161124
(1 row)

postgres=# explain  select x, x from (select (select random() offset 0) as x from (values(1),(2)) v(y)) ss;
                         QUERY PLAN                          
-------------------------------------------------------------
 Values Scan on "*VALUES*"  (cost=0.03..0.05 rows=2 width=0)
   InitPlan
     ->  Limit  (cost=0.00..0.01 rows=1 width=0)
           ->  Result  (cost=0.00..0.01 rows=1 width=0)
     ->  Limit  (cost=0.00..0.01 rows=1 width=0)
           ->  Result  (cost=0.00..0.01 rows=1 width=0)
(6 rows)
```

This was fixed in PostgreSQL 8.4, by commit 90b07dd7ba

**Comment by hsyuan** [comment]
Thanks, @hlinnaka.

**Comment by hlinnaka** [comment]
Status update on this: This bug still exists. Commit 90b07dd7ba, which fixed this in upstream, was made in 8.4.19. We're caught up to 8.4.0 in the merge now, so we still don't have that commit. We will merge through the minor releases only when we get closer to the GPDB6 release, in spring or next summer, and we'll get that fix then.

**Comment by hlinnaka** [comment]
We got the upstream fix for this in the 9.4 merge. So 'master' is now fixed. I don't think we have plans to change this in the stable releases, so I'm closing this issue. Thanks!
